src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -241,16 +377,33 @@ private static MethodHandle bindCaller(MethodHandle target, Class<?> hostClass)
      * @return true if the reference is not null and points to an object produced by {@code asInterfaceInstance}
      */
     public static boolean isWrapperInstance(Object x) {
-        return x instanceof WrapperInstance;
+        var clz = x.getClass();
+        var anno = clz.getDeclaredAnnotation(WrapperInstance.class);
+        if (anno == null) return false;
+
+        try {
+            return IMPLEMENTATION_INFOS.get(anno.implementedType()).resultClass == clz;
+        } catch (Throwable ex) {
+            return false; // Not an interface, etc.
+        }
     }
 
-    private static WrapperInstance asWrapperInstance(Object x) {
+    private static ImplementationInfo ensureWrapperInstance(Object x) {
+        var clz = x.getClass();
+        var anno = clz.getDeclaredAnnotation(WrapperInstance.class);
+        if (anno == null) throw newIllegalArgumentException("not a wrapper instance: " + x);;
+
+        ImplementationInfo ret;
         try {
-            if (x != null)
-                return (WrapperInstance) x;
-        } catch (ClassCastException ex) {
+            ret = IMPLEMENTATION_INFOS.get(anno.implementedType());
+        } catch (Throwable ex) {
```

Comment:
As mentioned by others, propagate `Error`s.

Reply:
This is already obsolete by https://github.com/liachmodded/jdk/commit/821d6b382aeceb70c67c59771deb7438f3b4b7fd, which I plan to push soon.